### PR TITLE
set resampler for sfx channels, remove alsoft config

### PR DESCRIPTION
**Pull request by fabiangreffrath**

### CODE REVIEW

**Comment by rfomin** [comment]
If we want this code, I suggest also removing the `alsoft` config file and set resampler for all sfx channels. GZDoom does this: https://github.com/ZDoom/gzdoom/blob/86bc9cafd49307a9367ba59b54c88d6525f73e69/src/common/audio/sound/oalsound.cpp#L848

**Comment by rfomin** [comment]
How about this:
<details>
 <summary>diff</summary>

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 1f35410..f614125 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -85,7 +85,6 @@ configure_file(config.h.in config.h)
 if(WIN32)
     install(FILES COPYING DESTINATION .)
     install(FILES README.md DESTINATION .)
-    install(FILES data/alsoft DESTINATION . RENAME alsoft.ini)
     install(DIRECTORY docs/ DESTINATION docs)
     install(DIRECTORY examples/ DESTINATION docs/examples)
     install(DIRECTORY autoload/ DESTINATION autoload)
@@ -95,7 +94,6 @@ if(WIN32)
 else()
     install(FILES COPYING DESTINATION "share/doc/${PROJECT_SHORTNAME}")
     install(FILES README.md DESTINATION "share/doc/${PROJECT_SHORTNAME}")
-    install(FILES data/alsoft DESTINATION "share/${PROJECT_SHORTNAME}" RENAME alsoft.conf)
     install(DIRECTORY autoload/ DESTINATION "share/${PROJECT_SHORTNAME}/autoload")
     if(EXISTS "${CMAKE_SOURCE_DIR}/docs")
         install(DIRECTORY docs/ DESTINATION "share/doc/${PROJECT_SHORTNAME}")
diff --git a/data/alsoft b/data/alsoft
deleted file mode 100644
index d64f8d4..0000000
--- a/data/alsoft
+++ /dev/null
@@ -1,3 +0,0 @@
-# OpenAL configuration. More options at
-# https://github.com/kcat/openal-soft/blob/master/alsoftrc.sample
-resampler = linear
diff --git a/src/i_sound.c b/src/i_sound.c
index 8dbbab1..e406f62 100644
--- a/src/i_sound.c
+++ b/src/i_sound.c
@@ -522,21 +522,22 @@ void I_ShutdownSound(void)
     snd_init = false;
 }
 
-/* C doesn't allow casting between function and non-function pointer types, so
- * with C99 we need to use a union to reinterpret the pointer type. Pre-C99
- * still needs to use a normal cast and live with the warning (C++ is fine with
- * a regular reinterpret_cast).
- */
+// C doesn't allow casting between function and non-function pointer types, so
+// with C99 we need to use a union to reinterpret the pointer type. Pre-C99
+// still needs to use a normal cast and live with the warning (C++ is fine with
+// a regular reinterpret_cast).
 #if __STDC_VERSION__ >= 199901L
 #define FUNCTION_CAST(T, ptr) (union{void *p; T f;}){ptr}.f
 #else
 #define FUNCTION_CAST(T, ptr) (T)(ptr)
 #endif
 
-static void I_PrintResamplerInfo(void)
+char *snd_resampler;
+
+static void SetResampler(void)
 {
-    LPALGETSTRINGISOFT alGetStringiSOFT;
-    ALint num_resamplers, def_resampler;
+    LPALGETSTRINGISOFT alGetStringiSOFT = NULL;
+    ALint i, num_resamplers, def_resampler;
 
     if (!alIsExtensionPresent("AL_SOFT_source_resampler"))
     {
@@ -546,18 +547,42 @@ static void I_PrintResamplerInfo(void)
 
     alGetStringiSOFT = FUNCTION_CAST(LPALGETSTRINGISOFT, alGetProcAddress("alGetStringiSOFT"));
 
+    if (!alGetStringiSOFT)
+    {
+        fprintf(stderr, "I_SetResampler: alGetStringiSOFT() is not avaible.\n");
+        return;
+    }
+
     num_resamplers = alGetInteger(AL_NUM_RESAMPLERS_SOFT);
     def_resampler = alGetInteger(AL_DEFAULT_RESAMPLER_SOFT);
 
     if (!num_resamplers)
     {
         printf(" No resamplers found!\n");
+        return;
     }
-    else
+
+    for (i = 0; i < num_resamplers; ++i)
+    {
+        if (!strcasecmp(snd_resampler, alGetStringiSOFT(AL_RESAMPLER_NAME_SOFT, i)))
+        {
+            def_resampler = i;
+            break;
+        }
+    }
+    if (i == num_resamplers)
+    {
+        printf(" Failed to find resampler: '%s'.\n", snd_resampler);
+        return;
+    }
+
+    for (i = 0; i < MAX_CHANNELS; ++i)
     {
-        const ALchar *name = alGetStringiSOFT(AL_RESAMPLER_NAME_SOFT, def_resampler);
-        printf(" Using %s resampler.\n", name);
+        alSourcei(openal_sources[i], AL_SOURCE_RESAMPLER_SOFT, def_resampler);
     }
+
+    printf(" Using '%s' resampler.\n",
+           alGetStringiSOFT(AL_RESAMPLER_NAME_SOFT, def_resampler));
 }
 
 // [FG] add links for likely missing sounds
@@ -627,11 +652,11 @@ void I_InitSound(void)
         return;
     }
 
-    I_PrintResamplerInfo();
-
     openal_sources = malloc(MAX_CHANNELS * sizeof(*openal_sources));
     alGenSources(MAX_CHANNELS, openal_sources);
 
+    SetResampler();
+
     I_AtExit(I_ShutdownSound, true);
 
     snd_init = true;
diff --git a/src/m_misc.c b/src/m_misc.c
index dc7dba2..e6ff1da 100644
--- a/src/m_misc.c
+++ b/src/m_misc.c
@@ -97,6 +97,7 @@ extern int winmm_chorus_level;
 #endif
 extern int opl_gain;
 extern int midi_player_menu;
+extern char *snd_resampler;
 extern boolean demobar;
 extern boolean smoothlight;
 extern boolean brightmaps;
@@ -406,6 +407,13 @@ default_t defaults[] = {
     "number of sound effects handled simultaneously"
   },
 
+  {
+    "snd_resampler",
+    (config_t *) &snd_resampler, NULL,
+    {.s = "linear"}, {0}, string, ss_gen, wad_no,
+    "OpenAL resampler (\"nearest\", \"linear\"(default), \"cubic\")"
+  },
+
   // [FG] music backend
   {
     "midi_player",
```
</details>

**Comment by fabiangreffrath** [comment]
> If we want this code, I suggest also removing the `alsoft` config file and set resampler for all sfx channels. GZDoom does this: https://github.com/ZDoom/gzdoom/blob/86bc9cafd49307a9367ba59b54c88d6525f73e69/src/common/audio/sound/oalsound.cpp#L848

The point is, I don't even *want* to use this code. The problem is just that you can't simply ask openal-soft "which resampler do you use". 😉

**Comment by fabiangreffrath** [comment]
> How about this:

Very nice!

**Comment by rfomin** [comment]
> The point is, I don't even _want_ to use this code. The problem is just that you can't simply ask openal-soft "which resampler do you use"

I think that just for printing it's not worth it. I like it if we get rid of `alsoft.ini` like in the patch in my previous post.

**Comment by fabiangreffrath** [comment]
We have been there before, right? 😁 

Could you please apply the change to this branch? That's easier than me copy/pasting it.

File: src/m_misc.c [hunk 1]
```diff
@@ -406,6 +407,13 @@ default_t defaults[] = {
     "number of sound effects handled simultaneously"
   },
 
+  {
+    "snd_resampler",
+    (config_t *) &snd_resampler, NULL,
+    {.s = "linear"}, {0}, string, ss_gen, wad_no,
+    "OpenAL resampler (\"nearest\", \"linear\"(default), \"cubic\")"
```
Comment (rfomin) [inline src/m_misc.c:414]: @fabiangreffrath OpenAL Soft report 7 resamplers on my system:
```
Nearest
Linear
Cubic
11th order Sinc (fast)
11th order Sinc
23rd order Sinc (fast)
23rd order Sinc
```
In order to properly present it, we need a menu option that we would populate dynamically. Although, I don't like another menu item for this.

Comment (fabiangreffrath) [reply src/m_misc.c:414]: I am fine with the first three, and also with not having this choice in the menu. The default is the same as before now, and the effort to change it, too.

Comment (fabiangreffrath) [reply src/m_misc.c:414]: Personally, I might even stick with cubic. It has a bit more "woom" than linear without sounding as scratchy. The sinc modes sound too different to me.

Comment (rfomin) [reply src/m_misc.c:414]: To be honest I don't hear much difference 😆

Comment (rfomin) [reply src/m_misc.c:414]: But what is closer to old Woof or Crispy sound? It's tempting to remove all of our printing and resampler settings, it just seems confusing to users, and OpenAL Soft knows best anyway.

Comment (mikeday0) [reply src/m_misc.c:414]: To me Crispy with `use_libsamplerate` set to `1`, latest Woof with `snd_resampler` set to `linear`, and Woof 10.5.1 all sound indistinguishable from one another.

Comment (fabiangreffrath) [reply src/m_misc.c:414]: > To me Crispy with `use_libsamplerate` set to `1`, latest Woof with `snd_resampler` set to `linear`, and Woof 10.5.1 all sound indistinguishable from one another.

Confirmed.

**Pull request merged**